simulator/func_sim/mars/mars_kernel.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * mars_kernel.cpp - MARS syscalls
+ * @author Vyacheslav Kompan kompan.vo@phystech.edu
+ * Copyright 2018 MIPT-MIPS
+ */
+
+#include "mars_kernel.h"
+/* Generic C++ */
+#include <string>
+
+static const uint8 v0 = 2;
+static const uint8 a0 = 4;
+
+bool MARSKernel::execute () {
+    uint64 syscall_code = sim.lock()->read_cpu_register( v0);
+    switch (syscall_code) {
+        case 1: print_integer(); break;
+        case 5: read_integer (); break;
+        case 10: return false; // exit
+        case 11: print_character(); break;
+        case 12: read_character(); break;
+        default: break;
+#if 0 // Temporarily, as some traces have unimplemented syscalls
+            throw Exception( "Syscall " + std::to_string( syscall_code) + " not implemented yet");
+#endif
+    }
+    return true;
+}
+
+void MARSKernel::print_integer() {
+    auto value = static_cast<int64>( sim.lock()->read_cpu_register( a0));
+    outstream << value;
+}
+
+void MARSKernel::read_integer() {
+    std::string input;
+    instream >> input;
+    uint64 value, pos;
+    try {
+        value = static_cast<uint64>( std::stoll( input, &pos, 0));
```

Comment:
It seems you have to use `boost/lexical_cast` to support 64-bit integers on 32-bit machines (`sizeof( long long) == 32`).